// memium/dist/array.js
import { withErrno } from 'kerium';
import { _throw } from 'utilium';
import { sizeof } from './misc.js';
import * as primitives from './primitives.js';
import { isType } from './types.js';
let _strictIndexes = false;
export function useStrictArrayIndexes() {
    _strictIndexes = true;
}
/**
 * The view on memory used for non-primitive array types.
 * This is a *value*
 */
export function StructArray(type, __length) {
    class StructArray extends DataView {
        length;
        type = type;
        *[Symbol.iterator]() {
            for (let i = 0; i < this.length; i++)
                yield this[i];
        }
        _offsets = [0];
        offsetOf(index) {
            if (!type.isDynamic)
                return index * type.size;
            if (index < this._offsets.length)
                return this._offsets[index];
            for (let i = this._offsets.length; i <= index; i++) {
                this._offsets[i] =
                    this._offsets[i - 1]
                        + sizeof(type.get(this.buffer, this.byteOffset + this._offsets[i - 1]));
            }
            return this._offsets[index];
        }
        constructor(lengthOrBuffer, byteOffset, byteLength) {
            const buffer = typeof lengthOrBuffer === 'object'
                ? lengthOrBuffer
                : new ArrayBuffer((lengthOrBuffer ?? 0) * type.size);
            super(buffer, byteOffset, byteLength);
            this.length =
                typeof lengthOrBuffer === 'number'
                    ? lengthOrBuffer
                    : type.isDynamic
                        ? (__length ?? _throw(`Unknown length of StructArray<${type.name}>`))
                        : Math.floor(this.byteLength / type.size);
            const offset = (i) => this.byteOffset + this.offsetOf(i);
            return new Proxy(this, {
                get(target, index) {
                    if (index in target)
                        return target[index];
                    const i = parseInt(index.toString());
                    if (!Number.isSafeInteger(i))
                        if (_strictIndexes)
                            throw withErrno('EINVAL', 'Invalid index: ' + index.toString());
                        else
                            return undefined;
                    return type.get(target.buffer, offset(i));
                },
                set(target, index, value) {
                    const i = parseInt(index.toString());
                    if (!Number.isSafeInteger(i))
                        if (_strictIndexes)
                            throw withErrno('EINVAL', 'Invalid index: ' + index.toString());
                        else
                            return false;
                    type.set(target.buffer, offset(i), value);
                    return true;
                },
            });
        }
    }
    for (const key of Object.getOwnPropertyNames(DataView.prototype)) {
        if (!key.startsWith('get') && !key.startsWith('set'))
            continue;
        Object.defineProperty(StructArray.prototype, key, {
            enumerable: false,
            configurable: false,
            writable: false,
            value: undefined,
        });
    }
    return StructArray;
}
function _isStructConstructor(arg) {
    return (typeof arg == 'function'
        && 'prototype' in arg
        && 'fields' in arg
        && typeof arg.fields == 'object'
        && isType(arg));
}
/**
 * A class used to create any *type* representing an array of a given "inner" or element type.
 */
export class ArrayType {
    type;
    length;
    name;
    size;
    __structArray;
    __arrayType;
    /** @internal @hidden */
    __isArrayType = true;
    constructor(type, length) {
        this.type = type;
        this.length = length;
        this.name = `${type.name}[${length}]`;
        this.size = type.size * length;
        this.array = StructArray(this);
        this.__structArray = StructArray(type, length);
        this.__arrayType = type.array ? type.array : this.__structArray;
    }
    get = (buffer, offset) => {
        if (primitives.isValid(this.type.name) && offset % this.type.size !== 0) {
            return new this.__structArray(buffer, offset, this.size);
        }
        return new this.__arrayType(buffer, offset, this.size);
    };
    set = (buffer, offset, value) => {
        if (this.length)
            for (let i = 0; i < this.length; i++) {
                this.type.set(buffer, offset + i * this.type.size, value[i]);
            }
        else {
            let pointer = offset;
            for (let i = 0; i < value.length; i++) {
                this.type.set(buffer, pointer, value[i]);
                pointer +=
                    _isStructConstructor(this.type) && this.type.isDynamic
                        ? sizeof(value[i])
                        : this.type.size;
            }
        }
    };
    /**
     * This is for an array of this array
     */
    array;
}
