// memium/dist/fields.internal.js
/* Internal stuff used for structs */
import { withErrno } from 'kerium';
import { ArrayType } from './array.js';
import { dynamicStructSize, isStructConstructor, isStructInstance, } from './structs.shared.js';
import { isType } from './types.js';
function _isBuilder(init) {
    return ((typeof init == 'object' || typeof init == 'function')
        && init !== null
        && 'toInit' in init
        && typeof init.toInit == 'function');
}
function _parseConfig(init) {
    if (_isBuilder(init))
        return init.toInit();
    if (isType(init))
        return { type: init };
    return init;
}
export function init(_name, init, extraOpts = {}) {
    if (!_name)
        throw withErrno('EINVAL', 'Invalid name for struct field');
    if (typeof _name == 'symbol')
        console.warn('Symbol used for struct field name will be coerced to string: ' + _name.toString());
    const name = _name.toString();
    const opt = Object.assign(_parseConfig(init), extraOpts);
    if (!isType(opt.type))
        throw withErrno('EINVAL', `Invalid type for struct field "${name}"`);
    const countedBy = !opt.countedBy
        ? ''
        : ` counted_by(${typeof opt.countedBy == 'string' ? opt.countedBy : '<function>'})`;
    return {
        name,
        offset: 0,
        type: opt.type,
        countedBy: opt.countedBy,
        alignment: opt.align ?? opt.type.size,
        decl: opt.type instanceof ArrayType
            ? `${opt.typeName ?? opt.type.type.name} ${name}[${opt.type.length}]${countedBy}`
            : `${opt.typeName ?? opt.type.name} ${name}`,
        littleEndian: !opt.bigEndian,
    };
}
function __fault(err, offset) {
    if (!(err instanceof Error) || err.message.toLowerCase() !== 'offset is outside the bounds of the dataview')
        throw err;
    const ex = withErrno('EFAULT', `Segmentation fault (at 0x${offset.toString(16)})`);
    Error.captureStackTrace(ex, __fault);
    throw ex;
}
export function isDynamicArray(instance, field) {
    return (field.type instanceof ArrayType
        && field.type.length == 0
        && !!field.countedBy?.length
        && !!instance.constructor.isDynamic);
}
function _count(instance, field) {
    let value = typeof field.countedBy == 'function'
        ? field.countedBy(instance)
        : instance[field.countedBy];
    if (typeof value == 'bigint') {
        if (value > BigInt(Number.MAX_SAFE_INTEGER)) {
            throw withErrno('EOVERFLOW', 'countedBy field value exceeds max safe integer');
        }
        value = Number(value);
    }
    if (typeof value != 'number')
        throw withErrno('EINVAL', 'countedBy field value is not a number');
    return value;
}
/**
 * Get the dynamic size in bytes of a dynamic array field
 * This *includes* the static size of the array items.
 */
export function dynamicArraySize(instance, field) {
    let size = field.type.type.size * _count(instance, field);
    if (isStructConstructor(field.type.type) && field.type.type.isDynamic) {
        for (const item of instance[field.name]) {
            size += dynamicStructSize(item);
        }
    }
    return size;
}
const kOffsets = Symbol('kOffsets');
/**
 * Get the offset of a field within a struct instance
 * @param cache If true, cache computed value and/or re-use existing cached value. If false, clear any cached offsets
 */
export function offsetOf(instance, targetField, cache) {
    let { offset, name } = targetField;
    instance[kOffsets] ||= Object.create(null);
    const offsetCache = instance[kOffsets];
    if (offsetCache[name] !== undefined) {
        if (cache)
            return offsetCache[name];
        else
            delete offsetCache[name];
    }
    const { fields } = instance.constructor;
    const index = fields.indexOf(targetField);
    for (const field of fields.slice(0, index)) {
        if (isDynamicArray(instance, field)) {
            offset += dynamicArraySize(instance, field);
        }
        const value = instance[field.name];
        if (isStructInstance(value) && value.constructor.isDynamic) {
            offset += dynamicStructSize(value);
        }
    }
    if (cache)
        offsetCache[name] = offset;
    else {
        for (const field of fields.slice(index))
            if (offsetCache[field.name] !== undefined)
                delete offsetCache[field.name];
    }
    return offset;
}
/** Sets the value of a field */
export function set(instance, field, value, index) {
    if (typeof value == 'string')
        value = value.charCodeAt(0);
    const offset = instance.byteOffset + offsetOf(instance, field, false) + (index ?? 0) * field.type.size;
    try {
        field.type.set(instance.buffer, offset, value);
        return;
    }
    catch (err) {
        __fault(err, offset);
    }
}
/** Gets the value of a field */
export function get(instance, field) {
    let type = field.type;
    if (isDynamicArray(instance, field)) {
        const inner = field.type.type;
        type = new ArrayType(inner, _count(instance, field));
    }
    const offset = instance.byteOffset + offsetOf(instance, field, true);
    try {
        return type.get(instance.buffer, offset);
    }
    catch (err) {
        __fault(err, offset);
    }
}
