// memium/dist/memory.js
/* eslint-disable @typescript-eslint/no-array-delete, @typescript-eslint/no-for-in-array */
// Work In Progress!
import { UV } from 'kerium';
import { BufferView } from 'utilium/buffer';
import { Pointer } from './pointer.js';
import { Void } from './types.js';
/**
 * Keeps track of the last `Memory` instance created.
 * This is used as defaults for various interfaces.
 * @internal
 */
export let _lastMemory;
/**
 * A memory manager that works with a single array buffer.
 * It is not designed for use-cases involving multiple realms or serialization.
 */
export class ArrayBufferMemory extends BufferView {
    get size() {
        return this.byteLength;
    }
    /**
     * A map of offsets to sections.
     * Spare arrays are used for auto-sort and other "magic".
     */
    map = [];
    constructor(buffer, byteOffset, byteLength) {
        super(buffer, byteOffset, byteLength);
        this.map.push({ size: this.byteLength, isFree: true });
        _lastMemory = this;
    }
    /**
     * Collects all free sections that are adjacent to the section at the given address.
     *
     * @remarks
     * This function assumes the provided section exists in the map.
     */
    collectFreeSections(at) {
        const sections = Object.entries(this.map).map(([k, s]) => [Number(k), s]);
        const i = sections.findIndex(([off]) => off === at);
        if (i === -1)
            throw UV('EINVAL');
        let primary = sections[i];
        // "left" side
        for (const [off, section] of sections.slice(0, i).reverse()) {
            if (!section.isFree)
                break;
            section.size += primary[1].size;
            delete this.map[primary[0]];
            primary = [off, section];
        }
        // "right" side
        for (const [off, section] of sections.slice(i + 1)) {
            if (!section.isFree)
                break;
            primary[1].size += section.size;
            delete this.map[off];
        }
    }
    alloc(size) {
        if (size > this.byteLength)
            throw UV('ENOMEM', 'alloc');
        for (const key in this.map) {
            const off = Number(key);
            const { size: sectionSize, isFree } = this.map[off];
            if (!isFree || sectionSize < size)
                continue;
            this.map[off + size] = { size: sectionSize - size, isFree: true };
            this.map[off] = { size, isFree: false };
            return new Pointer(Void, off, this);
        }
        throw UV('ENOMEM', 'alloc');
    }
    free(addr) {
        addr = addr.valueOf();
        if (!(addr in this.map))
            throw UV('EINVAL', 'free');
        this.map[addr].isFree = true;
        queueMicrotask(() => this.collectFreeSections(addr));
    }
    realloc(at, size) {
        const pointer = at instanceof Pointer ? at : new Pointer(Void, at, this);
        const addr = at instanceof Pointer ? pointer.valueOf() : at;
        if (addr > this.byteLength)
            throw UV('EFAULT');
        if (size > this.byteLength)
            throw UV('ENOMEM');
        if (!(addr in this.map))
            throw UV('EINVAL', 'realloc');
        queueMicrotask(() => this.collectFreeSections(addr));
        const off = Number(addr);
        const oldSize = this.map[off].size;
        if (oldSize >= size) {
            this.map[off].size = size;
            this.map[off + size] = { size: oldSize - size, isFree: true };
            return pointer;
        }
        let needsMove = false;
        const free = [];
        for (const key in this.map) {
            const off = Number(key);
            if (off <= addr + oldSize && off >= addr + size)
                continue;
            if (this.map[off].isFree)
                free.push(off);
            else
                needsMove = true;
        }
        if (!needsMove) {
            for (const off of free)
                delete this.map[off];
            this.map[addr].size = size;
            return pointer;
        }
        const view = new Uint8Array(this.buffer);
        const newPointer = new Pointer(pointer.type, this.alloc(size).valueOf(), this);
        view.copyWithin(newPointer.valueOf(), addr, addr + oldSize);
        this.map[addr].isFree = true;
        return newPointer;
    }
    usage() {
        let used = 0, free = 0;
        for (const key in this.map) {
            const off = Number(key);
            const { size, isFree } = this.map[off];
            if (isFree)
                free += size;
            else
                used += size;
        }
        return { total: this.byteLength, used, free };
    }
    at(addr) {
        if (addr > this.byteLength)
            throw UV('EFAULT');
        return new Uint8Array(this.buffer, addr, this.byteLength);
    }
}
